src/consensus.rs

```diff
@@ -0,0 +1,307 @@
+use log::info;
+use rand::{CryptoRng, Rng};
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::{Error, Result};
+use crate::{PublicKeyShare, SecretKeyShare};
+
+use crate::vote::{Ballot, Proposition, UnsignedVote, Vote};
+
+// TODO tmp to separate files
+use crate::membership::{MembershipState, Reconfig};
+
+#[derive(Debug)]
+pub struct State<T: Proposition> {
+    pub secret_key: SecretKeyShare,
+    pub elders: BTreeSet<PublicKeyShare>,
+    pub votes: BTreeMap<PublicKeyShare, Vote<T>>,
+}
```

Comment:
This partial refactor is going to be too difficult to work with, sorry. Lets just start over and do a simple file split, and then seperate PR's for each change.

Breaking apart sn_membership.rs will be a clear improvement over the status-quo, where-as this approach I fear will be too complex to manage.

Reply:
I renamed the `State` to `MembershipState`, all the methods are in there (and untouched), this `State` is just a sub-struct (maybe I should have named it differently?)
What we have here is similar as before: 
```rust
#[derive(Debug)]
pub struct MembershipState<T: Proposition> {
    pub state: State {
        pub secret_key: SecretKeyShare,
        pub elders: BTreeSet<PublicKeyShare>,
        pub votes: BTreeMap<PublicKeyShare, Vote<T>>,
    },
    pub gen: Generation,
    pub pending_gen: Generation,
    pub forced_reconfigs: BTreeMap<Generation, BTreeSet<Reconfig<T>>>, // TODO: change to bootstrap members
    pub history: BTreeMap<Generation, Vote<T>>, // for onboarding new procs, the vote proving super majority
}
```
It's the same as before but with one more level of dereferencing. 

Or did I miss your point? In that case, could you explain why this "is going to be too difficult to work with"?

Reply:
I don't yet fully buy that we need to move `gen` / `forced_reconfigs` / `history` out of a core consensus algorithm, so doing an early structure split like this will just create churn if this is not needed / not wanted.

Once we do separate out the two structs, I think it's important to be able to see the changes on their own to make sure we're not introducing regressions.

Besides that, I think you're missing some changes introduced on master / your changing master at the same time and it's hard to see the what exactly is changing (`Name` is not Copy on master, but `Proposition` is).

I just want to make sure we're not losing anything in the shuffle.

> could you explain why this "is going to be too difficult to work with"?
Sorry, I should have just explained myself better in my first comment, rather than vague platitudes, my bad.